In [2]:
from DataAugmentation import data_loader
import os
root_data_dir = r"D:\Projects\MulticlassFishImageClassification\Project_data\data"

train_dir = os.path.join(root_data_dir, 'train')
valid_dir = os.path.join(root_data_dir, 'val')
test_dir = os.path.join(root_data_dir, 'test')

train_generator, valid_generator, test_generator = data_loader(train_dir,valid_dir,test_dir)


Found 6225 images belonging to 11 classes.
Found 1092 images belonging to 11 classes.
Found 3187 images belonging to 11 classes.


### MODEL Traing Pipeline


In [3]:
import tensorflow as tf
from tensorflow.keras import layers, models
from tensorflow.keras.callbacks import EarlyStopping, ModelCheckpoint
from tensorflow.keras.applications import (
    VGG16, ResNet50, MobileNetV2, InceptionV3, EfficientNetB0
)

NUM_CLASSES = train_generator.num_classes
IMG_SIZE = (256, 256)
EPOCHS = 15

callbacks = [
    EarlyStopping(monitor='val_accuracy', patience=5, restore_best_weights=True),
    ModelCheckpoint("best_model.h5", monitor="val_accuracy", save_best_only=True)
]


### CNN Model From Scratch

In [10]:
def build_cnn(num_classes):
    model = tf.keras.Sequential([
        layers.Input(shape=(256,256,3)),
        
        layers.Conv2D(32, 3, activation="relu"),
        layers.MaxPooling2D(),

        layers.Conv2D(64, 3, activation="relu"),
        layers.MaxPooling2D(),

        layers.Conv2D(128, 3, activation="relu"),
        layers.MaxPooling2D(),

        layers.Flatten(),
        layers.Dense(256, activation="relu"),
        layers.Dropout(0.5),
        layers.Dense(num_classes, activation="softmax")
    ])

    model.compile(optimizer="adam", loss="categorical_crossentropy", metrics=["accuracy"])
    return model


In [4]:
from tensorflow.keras.models import Model


### VGG16 Model From Scratch

In [18]:
def build_vgg16(num_classes):
    base = VGG16(include_top=False, weights="imagenet", input_shape=(256,256,3))
    base.trainable = False

    x = base.output
    x = layers.GlobalAveragePooling2D()(x)
    x = layers.Dense(256, activation="relu")(x)  # new layer instance
    x = layers.Dropout(0.5)(x)
    
    output = layers.Dense(num_classes, activation="softmax")(x)

    model = Model(inputs=base.input, outputs=output)
    model.compile(optimizer="adam", loss="categorical_crossentropy", metrics=["accuracy"])
    return model


### ResNet50 Model From Scratch

In [5]:
def build_resnet50(num_classes):
    base = ResNet50(include_top=False, weights="imagenet", input_shape=(256,256,3))
    base.trainable = False

    x = base.output
    x = layers.GlobalAveragePooling2D()(x)
    x = layers.Dense(256, activation="relu")(x)
    x = layers.Dropout(0.5)(x)

    output = layers.Dense(num_classes, activation="softmax")(x)

    model = Model(inputs=base.input, outputs=output)
    model.compile(optimizer="adam", loss="categorical_crossentropy", metrics=["accuracy"])
    return model


### MobileNetV2 Model From Scratch

In [9]:
def build_mobilenet(num_classes):
    base = MobileNetV2(include_top=False, weights="imagenet", input_shape=(256,256,3))
    base.trainable = False

    x = base.output
    x = layers.GlobalAveragePooling2D()(x)
    x = layers.Dense(256, activation="relu")(x)
    x = layers.Dropout(0.5)(x)

    output = layers.Dense(num_classes, activation="softmax")(x)

    model = Model(inputs=base.input, outputs=output)
    model.compile(optimizer="adam", loss="categorical_crossentropy", metrics=["accuracy"])
    return model


### InceptionV3 Model From Scratch

In [8]:
def build_inceptionv3(num_classes):
    base = InceptionV3(include_top=False, weights="imagenet", input_shape=(256,256,3))
    base.trainable = False

    x = base.output
    x = layers.GlobalAveragePooling2D()(x)
    x = layers.Dense(256, activation="relu")(x)
    x = layers.Dropout(0.5)(x)

    output = layers.Dense(num_classes, activation="softmax")(x)

    model = Model(inputs=base.input, outputs=output)
    model.compile(optimizer="adam", loss="categorical_crossentropy", metrics=["accuracy"])
    return model


### Train CNN Model

In [15]:
model = build_cnn(NUM_CLASSES)
history = model.fit(train_generator, validation_data=valid_generator, epochs=10)
model.save("D:/Projects/MulticlassFishImageClassification/Models/CNN_model.h5")


d:\Projects\MulticlassFishImageClassification\env\Lib\site-packages\keras\src\trainers\data_adapters\py_dataset_adapter.py:121: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


Epoch 1/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 421s 2s/step - accuracy: 0.4071 - loss: 1.6979 - val_accuracy: 0.6392 - val_loss: 1.0926
Epoch 2/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 313s 2s/step - accuracy: 0.6254 - loss: 1.0461 - val_accuracy: 0.7866 - val_loss: 0.6489
Epoch 3/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 296s 2s/step - accuracy: 0.7187 - loss: 0.8303 - val_accuracy: 0.8397 - val_loss: 0.5011
Epoch 4/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 304s 2s/step - accuracy: 0.7804 - loss: 0.6294 - val_accuracy: 0.8397 - val_loss: 0.5299
Epoch 5/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 321s 2s/step - accuracy: 0.8117 - loss: 0.5414 - val_accuracy: 0.9130 - val_loss: 0.3026
Epoch 6/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 297s 2s/step - accuracy: 0.8320 - loss: 0.4930 - val_accuracy: 0.9322 - val_loss: 0.2806
Epoch 7/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 291s 1s/step - accuracy: 0.8564 - loss: 0.4162 - val_accuracy: 0.9396 - val_loss: 0.2063
Epoch 8/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 297s 2s/step - accuracy: 0.8696 - loss: 0.3732 - val_accu

### Train VGG16 Model

In [22]:
model_vgg16 = build_vgg16(NUM_CLASSES)
history = model_vgg16.fit(train_generator, validation_data=valid_generator, epochs=10)
model_vgg16.save("D:/Projects/MulticlassFishImageClassification/Models/VGG16_model.h5")


Epoch 1/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 2320s 12s/step - accuracy: 0.4606 - loss: 1.6647 - val_accuracy: 0.7408 - val_loss: 1.1246
Epoch 2/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 2062s 11s/step - accuracy: 0.7080 - loss: 0.9801 - val_accuracy: 0.8590 - val_loss: 0.6661
Epoch 3/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 2477s 13s/step - accuracy: 0.8032 - loss: 0.6803 - val_accuracy: 0.9158 - val_loss: 0.4640
Epoch 4/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 1223s 6s/step - accuracy: 0.8361 - loss: 0.5511 - val_accuracy: 0.9112 - val_loss: 0.3832
Epoch 5/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 1425s 7s/step - accuracy: 0.8628 - loss: 0.4526 - val_accuracy: 0.9414 - val_loss: 0.2830
Epoch 6/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 1784s 9s/step - accuracy: 0.8843 - loss: 0.3942 - val_accuracy: 0.9304 - val_loss: 0.2755
Epoch 7/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 1217s 6s/step - accuracy: 0.8964 - loss: 0.3447 - val_accuracy: 0.9469 - val_loss: 0.2216
Epoch 8/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 2670s 14s/step - accuracy: 0.9102 - loss: 0.311

### Train ResNet50 Model

In [6]:
model_resnet50 = build_resnet50(NUM_CLASSES)
history = model_resnet50.fit(train_generator, validation_data=valid_generator, epochs=10)
model_resnet50.save("D:/Projects/MulticlassFishImageClassification/Models/ResNet50_model.h5")


94765736/94765736 ━━━━━━━━━━━━━━━━━━━━ 20s 0us/step


d:\Projects\MulticlassFishImageClassification\env\Lib\site-packages\keras\src\trainers\data_adapters\py_dataset_adapter.py:121: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


Epoch 1/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 789s 4s/step - accuracy: 0.1647 - loss: 2.3318 - val_accuracy: 0.1712 - val_loss: 2.2365
Epoch 2/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 856s 4s/step - accuracy: 0.1841 - loss: 2.2273 - val_accuracy: 0.1712 - val_loss: 2.1801
Epoch 3/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 622s 3s/step - accuracy: 0.1942 - loss: 2.1666 - val_accuracy: 0.2051 - val_loss: 2.1007
Epoch 4/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 607s 3s/step - accuracy: 0.2077 - loss: 2.1318 - val_accuracy: 0.2262 - val_loss: 2.0689
Epoch 5/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 607s 3s/step - accuracy: 0.2218 - loss: 2.1087 - val_accuracy: 0.2427 - val_loss: 2.0304
Epoch 6/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 974s 5s/step - accuracy: 0.2350 - loss: 2.0843 - val_accuracy: 0.3132 - val_loss: 1.9859
Epoch 7/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 650s 3s/step - accuracy: 0.2384 - loss: 2.0599 - val_accuracy: 0.2674 - val_loss: 1.9883
Epoch 8/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 617s 3s/step - accuracy: 0.2540 - loss: 2.0396 - val_accu

### Train MobileNet Model

In [ ]:
model_mobilenet = build_mobilenet(NUM_CLASSES)
history = model_mobilenet.fit(train_generator, validation_data=valid_generator, epochs=10)
model_mobilenet.save("D:/Projects/MulticlassFishImageClassification/Models/MobileNetV2_model.h5")


C:\Users\arune\AppData\Local\Temp\ipykernel_15196\4281020698.py:2: UserWarning: `input_shape` is undefined or non-square, or `rows` is not in [96, 128, 160, 192, 224]. Weights for input shape (224, 224) will be loaded as the default.
  base = MobileNetV2(include_top=False, weights="imagenet", input_shape=(256,256,3))


9406464/9406464 ━━━━━━━━━━━━━━━━━━━━ 9s 1us/step
Epoch 1/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 324s 2s/step - accuracy: 0.8476 - loss: 0.5088 - val_accuracy: 0.9734 - val_loss: 0.0939
Epoch 2/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 329s 2s/step - accuracy: 0.9589 - loss: 0.1359 - val_accuracy: 0.9872 - val_loss: 0.0420
Epoch 3/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 6128s 32s/step - accuracy: 0.9680 - loss: 0.0999 - val_accuracy: 0.9872 - val_loss: 0.0381
Epoch 4/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 316s 2s/step - accuracy: 0.9733 - loss: 0.0827 - val_accuracy: 0.9908 - val_loss: 0.0289
Epoch 5/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 356s 2s/step - accuracy: 0.9790 - loss: 0.0651 - val_accuracy: 0.9927 - val_loss: 0.0223
Epoch 6/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 373s 2s/step - accuracy: 0.9807 - loss: 0.0603 - val_accuracy: 0.9963 - val_loss: 0.0156
Epoch 7/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 431s 2s/step - accuracy: 0.9843 - loss: 0.0494 - val_accuracy: 0.9908 - val_loss: 0.0239
Epoch 8/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 452s 2

### Train InceptionV3 Model

In [12]:
model_inception = build_inceptionv3(NUM_CLASSES)
history = model_inception.fit(train_generator, validation_data=valid_generator, epochs=10)
model_inception.save("D:/Projects/MulticlassFishImageClassification/Models/InceptionV300_model.h5")


87910968/87910968 ━━━━━━━━━━━━━━━━━━━━ 17s 0us/step
Epoch 1/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 383s 2s/step - accuracy: 0.7685 - loss: 0.6802 - val_accuracy: 0.9441 - val_loss: 0.1994
Epoch 2/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 367s 2s/step - accuracy: 0.9059 - loss: 0.2922 - val_accuracy: 0.9478 - val_loss: 0.1353
Epoch 3/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 365s 2s/step - accuracy: 0.9200 - loss: 0.2404 - val_accuracy: 0.9597 - val_loss: 0.1120
Epoch 4/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 364s 2s/step - accuracy: 0.9391 - loss: 0.1853 - val_accuracy: 0.9606 - val_loss: 0.1041
Epoch 5/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 364s 2s/step - accuracy: 0.9345 - loss: 0.1879 - val_accuracy: 0.9606 - val_loss: 0.1038
Epoch 6/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 365s 2s/step - accuracy: 0.9396 - loss: 0.1697 - val_accuracy: 0.9762 - val_loss: 0.0710
Epoch 7/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 365s 2s/step - accuracy: 0.9499 - loss: 0.1441 - val_accuracy: 0.9817 - val_loss: 0.0576
Epoch 8/10
195/195 ━━━━━━━━━━━━━━━━━━━━ 368s 

In [14]:
train_generator.class_indices


{'animal fish': 0,
 'animal fish bass': 1,
 'fish sea_food black_sea_sprat': 2,
 'fish sea_food gilt_head_bream': 3,
 'fish sea_food hourse_mackerel': 4,
 'fish sea_food red_mullet': 5,
 'fish sea_food red_sea_bream': 6,
 'fish sea_food sea_bass': 7,
 'fish sea_food shrimp': 8,
 'fish sea_food striped_red_mullet': 9,
 'fish sea_food trout': 10}

### Evaluate all trained Model

In [15]:
from sklearn.metrics import classification_report, confusion_matrix
import numpy as np

def evaluate_model(model, name):
    print(f"Evaluating {name}...")

    loss, acc = model.evaluate(test_generator)
    preds = model.predict(test_generator)
    y_pred = np.argmax(preds, axis=1)
    y_true = test_generator.classes

    print("\nAccuracy:", acc)
    print("\nClassification Report:")
    print(classification_report(y_true, y_pred))

    return acc


In [16]:
results = {}

for model_name in ["CNN", "VGG16", "ResNet50", "MobileNetV2", "InceptionV300"]:
    model = tf.keras.models.load_model(f"D:/Projects/MulticlassFishImageClassification/Models/{model_name}_model.h5")
    acc = evaluate_model(model, model_name)
    results[model_name] = acc

best_model = max(results, key=results.get)
best_model, results[best_model]


d:\Projects\MulticlassFishImageClassification\env\Lib\site-packages\keras\src\trainers\data_adapters\py_dataset_adapter.py:121: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


Evaluating CNN...
100/100 ━━━━━━━━━━━━━━━━━━━━ 77s 766ms/step - accuracy: 0.9247 - loss: 0.2199
100/100 ━━━━━━━━━━━━━━━━━━━━ 32s 314ms/step

Accuracy: 0.9246940612792969

Classification Report:
              precision    recall  f1-score   support

           0       0.96      1.00      0.98       520
           1       0.00      0.00      0.00        13
           2       0.98      0.97      0.98       298
           3       0.80      0.85      0.82       305
           4       1.00      0.95      0.98       286
           5       0.98      0.98      0.98       291
           6       0.99      0.65      0.78       273
           7       0.75      0.98      0.85       327
           8       1.00      0.93      0.97       289
           9       0.93      0.89      0.91       293
          10       0.94      1.00      0.97       292

    accuracy                           0.92      3187
   macro avg       0.85      0.84      0.84      3187
weighted avg       0.93      0.92      0.92     

d:\Projects\MulticlassFishImageClassification\env\Lib\site-packages\sklearn\metrics\_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
d:\Projects\MulticlassFishImageClassification\env\Lib\site-packages\sklearn\metrics\_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
d:\Projects\MulticlassFishImageClassification\env\Lib\site-packages\sklearn\metrics\_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier,

Evaluating VGG16...
100/100 ━━━━━━━━━━━━━━━━━━━━ 790s 8s/step - accuracy: 0.9699 - loss: 0.1391
100/100 ━━━━━━━━━━━━━━━━━━━━ 987s 10s/step

Accuracy: 0.9698776006698608

Classification Report:
              precision    recall  f1-score   support

           0       0.98      0.99      0.98       520
           1       0.00      0.00      0.00        13
           2       0.98      0.98      0.98       298
           3       0.98      0.96      0.97       305
           4       0.93      1.00      0.96       286
           5       0.96      0.93      0.95       291
           6       0.99      0.99      0.99       273
           7       0.98      0.99      0.98       327
           8       0.95      1.00      0.98       289
           9       0.94      0.90      0.92       293
          10       1.00      0.98      0.99       292

    accuracy                           0.97      3187
   macro avg       0.88      0.88      0.88      3187
weighted avg       0.97      0.97      0.97      

d:\Projects\MulticlassFishImageClassification\env\Lib\site-packages\sklearn\metrics\_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
d:\Projects\MulticlassFishImageClassification\env\Lib\site-packages\sklearn\metrics\_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
d:\Projects\MulticlassFishImageClassification\env\Lib\site-packages\sklearn\metrics\_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier,

Evaluating ResNet50...
100/100 ━━━━━━━━━━━━━━━━━━━━ 419s 4s/step - accuracy: 0.3037 - loss: 1.9679
100/100 ━━━━━━━━━━━━━━━━━━━━ 422s 4s/step

Accuracy: 0.3037339150905609

Classification Report:
              precision    recall  f1-score   support

           0       0.56      0.65      0.61       520
           1       0.00      0.00      0.00        13
           2       1.00      0.00      0.01       298
           3       0.00      0.00      0.00       305
           4       0.19      0.69      0.30       286
           5       0.28      0.02      0.03       291
           6       0.34      0.11      0.17       273
           7       0.68      0.04      0.08       327
           8       0.24      0.58      0.34       289
           9       0.32      0.13      0.18       293
          10       0.29      0.61      0.39       292

    accuracy                           0.30      3187
   macro avg       0.36      0.26      0.19      3187
weighted avg       0.41      0.30      0.23    

d:\Projects\MulticlassFishImageClassification\env\Lib\site-packages\sklearn\metrics\_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
d:\Projects\MulticlassFishImageClassification\env\Lib\site-packages\sklearn\metrics\_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
d:\Projects\MulticlassFishImageClassification\env\Lib\site-packages\sklearn\metrics\_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier,

Evaluating MobileNetV2...
100/100 ━━━━━━━━━━━━━━━━━━━━ 135s 1s/step - accuracy: 0.9937 - loss: 0.0208
100/100 ━━━━━━━━━━━━━━━━━━━━ 136s 1s/step

Accuracy: 0.993724524974823

Classification Report:
              precision    recall  f1-score   support

           0       0.98      1.00      0.99       520
           1       1.00      0.23      0.38        13
           2       1.00      0.99      1.00       298
           3       1.00      1.00      1.00       305
           4       1.00      1.00      1.00       286
           5       0.99      1.00      0.99       291
           6       1.00      1.00      1.00       273
           7       0.99      1.00      0.99       327
           8       1.00      1.00      1.00       289
           9       1.00      0.99      0.99       293
          10       1.00      1.00      1.00       292

    accuracy                           0.99      3187
   macro avg       1.00      0.93      0.94      3187
weighted avg       0.99      0.99      0.99  

Evaluating InceptionV300...
100/100 ━━━━━━━━━━━━━━━━━━━━ 266s 3s/step - accuracy: 0.9878 - loss: 0.0395
100/100 ━━━━━━━━━━━━━━━━━━━━ 270s 3s/step

Accuracy: 0.9877628087997437

Classification Report:
              precision    recall  f1-score   support

           0       0.98      0.99      0.99       520
           1       1.00      0.23      0.38        13
           2       0.99      0.99      0.99       298
           3       0.97      1.00      0.99       305
           4       0.99      1.00      0.99       286
           5       0.99      1.00      1.00       291
           6       0.99      0.99      0.99       273
           7       0.98      0.99      0.98       327
           8       1.00      1.00      1.00       289
           9       1.00      0.96      0.98       293
          10       1.00      0.98      0.99       292

    accuracy                           0.99      3187
   macro avg       0.99      0.92      0.93      3187
weighted avg       0.99      0.99      0.9

('MobileNetV2', 0.993724524974823)

### Accuracy Comparison
#### Best performing models (Accuracy):
##### MobileNetV2 – 99.37%
##### InceptionV3 – 98.77%
##### VGG16 – 96.99%
##### CNN – 92.46%
##### ResNet50 – 30.37% (Failed Model)

#### MobileNetV2 Model is Best Overall since It gives Best accuracy, Best overall precision/recall/f1 across classes and it is Lightweight and fast

### Model Comparision Report

| Model           | Accuracy   | Macro Precision | Macro Recall | Macro F1-score |
| --------------- | ---------- | --------------- | ------------ | -------------- |
| **CNN**         | 0.9247     | 0.85            | 0.84         | 0.84           |
| **VGG16**       | 0.9699     | 0.88            | 0.88         | 0.88           |
| **ResNet50**    | 0.3037     | 0.36            | 0.26         | 0.19           |
| **MobileNetV2** | **0.9937** | **1.00**        | 0.93         | 0.94           |
| **InceptionV3** | 0.9878     | 0.99            | 0.92         | 0.93           |
